### PR TITLE
HAAR-2001: add hidden role and endpoint for adding hidden role for a …

### PR DESCRIPTION
**Pull request by SimonMitchellMOJ**
…service, for IMS

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleController.kt [hunk 1]
```diff
@@ -149,13 +151,18 @@ class UserRoleController(
     @PathVariable
     role: String,
   ) {
-    userRoleService.removeRoleByUserId(userId, role)
+    if (authenticationFacade.getAuthentication().authorities.map { it.authority }.any { it == "ROLE_MAINTAIN_IMS_USERS" }
+    ) {
+      userRoleService.serviceRemoveRoleByUserId(userId, role)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleController.kt:156]: bit of a naff name for a service method - doesn't distinguish between that and the one below.  Would be good to add some javadoc to service methods too explaining which to call and when

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserRoleService.kt [hunk 2]
```diff
@@ -74,19 +78,29 @@ class UserRoleService(
...
+      }
+
       val formattedRoles = roleCodes.map { formatRole(it) }
-      val allAssignableRoles = getAllAssignableRolesByUserId(userId)
       for (roleCode in formattedRoles) {
         // check that role exists
         val role = roleRepository.findByRoleCode(roleCode) ?: throw UserRoleException("role", "role.notfound")
 
-        val userRoles = getUserRoles(userId) ?: throw NotFoundException("usernotfound")
+        val userRoles = roleRepository.findRolesByUserId(userId).toList()
```
Comment (SimonMitchellMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserRoleService.kt:93]: getUserRoles was doing another maintainUserCheck which is not needed so just getting the user roles now

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleController.kt (hunk 1 above)
Comment (SimonMitchellMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleController.kt]: yep - removed service method!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleControllerTest.kt [hunk 3]
```diff
@@ -59,6 +59,8 @@ class UserRoleControllerTest {
 
   @Test
   fun addRolesByUserId(): Unit = runBlocking {
+    // whenever(authenticationFacade.getAuthentication()).thenReturn(authentication)
```
Comment (ravmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleControllerTest.kt:62]: remove commented lines

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleControllerTest.kt [hunk 4]
```diff
@@ -68,15 +70,23 @@ class UserRoleControllerTest {
 
   @Test
   fun removeRoleByUserId_success(): Unit = runBlocking {
-    val roleId = UUID.randomUUID()
-    userRoleController.removeRoleByUserId(roleId, "roleCode")
-    verify(userRoleService).removeRoleByUserId(roleId, "roleCode")
+    // whenever(authenticationFacade.getAuthentication()).thenReturn(authentication)
+    // whenever(authentication.authorities).thenReturn(listOf(SimpleGrantedAuthority("ROLE_MAINTAIN_OAUTH_USERS")))
```
Comment (ravmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserRoleControllerTest.kt:74]: remove commented lines

**Review by ravmoj — APPROVED** [review]
lgtm

**Pull request merged**